src/config_file.c

```diff
@@ -1568,11 +1571,148 @@ static int config_parse(
 }
 
 struct parse_data {
+	const git_repository *repo;
+	const char *file_path;
 	git_strmap *values;
 	git_config_level_t level;
 	int depth;
 };
 
+static int parse_include(struct reader *reader,
+	struct parse_data *parse_data, const char *file)
+{
+	struct config_file *include;
+	git_buf path = GIT_BUF_INIT;
+	char *dir;
+	int result;
+
+	if ((result = git_path_dirname_r(&path, reader->file->path)) < 0)
+		return result;
+
+	dir = git_buf_detach(&path);
+	result = included_path(&path, dir, file);
+	git__free(dir);
+
+	if (result < 0)
+		return result;
+
+	include = git_array_alloc(reader->file->includes);
+	memset(include, 0, sizeof(*include));
+	git_array_init(include->includes);
+	include->path = git_buf_detach(&path);
+
+	result = config_read(parse_data->repo, parse_data->values,
+		include, parse_data->level, parse_data->depth+1);
+
+	if (result == GIT_ENOTFOUND) {
+		giterr_clear();
+		result = 0;
+	}
+
+	return result;
+}
+
+static int do_match_gitdir(
+	int *matches,
+	const git_repository *repo,
+	const char *cfg_file,
+	const char *value,
+	bool case_insensitive)
+{
+	git_buf path = GIT_BUF_INIT;
+	int error, fnmatch_flags;
+
+	if (value[0] == '.' && git_path_is_dirsep(value[1])) {
+		git_path_dirname_r(&path, cfg_file);
+		git_buf_joinpath(&path, path.ptr, value + 2);
+	} else if (value[0] == '~' && git_path_is_dirsep(value[1]))
+		git_sysdir_expand_global_file(&path, value + 1);
+	else if (!git_path_is_absolute(value))
+		git_buf_joinpath(&path, "**", value);
+	else
+		git_buf_sets(&path, value);
+
+	if (git_buf_oom(&path)) {
+		error = -1;
+		goto out;
+	}
+
+	if (git_path_is_dirsep(value[strlen(value) - 1]))
+		git_buf_puts(&path, "**");
+
+	fnmatch_flags = FNM_PATHNAME|FNM_LEADING_DIR;
+	if (case_insensitive)
+		fnmatch_flags |= FNM_IGNORECASE;
+
+	if ((error = p_fnmatch(path.ptr, git_repository_path(repo), fnmatch_flags)) < 0)
+		goto out;
+
+	*matches = (error == 0);
+
+out:
+	git_buf_free(&path);
+	return error;
+}
+
+static int conditional_match_gitdir(
+	int *matches,
+	const git_repository *repo,
+	const char *cfg_file,
+	const char *value)
+{
+	return do_match_gitdir(matches, repo, cfg_file, value, false);
+}
+
+static int conditional_match_gitdir_i(
+	int *matches,
+	const git_repository *repo,
+	const char *cfg_file,
+	const char *value)
+{
+	return do_match_gitdir(matches, repo, cfg_file, value, true);
+}
+
+static const struct {
+	const char *prefix;
+	int (*matches)(int *matches, const git_repository *repo, const char *cfg, const char *value);
+} conditions[] = {
+	{ "gitdir:", conditional_match_gitdir },
+	{ "gitdir/i:", conditional_match_gitdir_i }
+};
+
+static int parse_conditional_include(struct reader *reader,
+	struct parse_data *parse_data, const char *section, const char *file)
+{
+	char *condition;
+	size_t i;
+	int error = 0, matches;
+
+	if (!parse_data->repo)
+		return 0;
+
+	condition = git__substrdup(section + strlen("includeIf."),
+				   strlen(section) - strlen("includeIf.") - strlen(".path"));
```

Comment:
This might just be me, but I prefer the method used in the normalisation which searches for the first and last `'.'` to figure out how long the middle part is without making us count hard-coded strings.